state/state.go

```diff
@@ -1118,6 +1118,47 @@ var (
 	errLocalApplicationExists          = errors.Errorf("application already exists")
 )
 
+type AddCloudServiceArgs struct {
+	// Id will be the application Name if it's a part of application,
+	// and will be controller UUID for k8s a controller(controller does Not have an application).
+	Id         string
+	ProviderId string
+	Addresses  []network.Address
+}
+
+// AddCloudService creates a cloud service.
+func (st *State) AddCloudService(args AddCloudServiceArgs) (_ *CloudService, err error) {
+	defer errors.DeferredAnnotatef(&err, "cannot add cloud service %q", args.ProviderId)
+
+	doc := cloudServiceDoc{
+		DocID:      args.Id,
+		ProviderId: args.ProviderId,
+		Addresses:  fromNetworkAddresses(args.Addresses, OriginProvider),
+	}
+	svc := newCloudService(st, &doc)
+
+	ops, err := svc.saveServiceOps(doc)
```

Comment:
This needs to be part of a buildTxn loop, to cater for insert vs update races.
See for example (ec *externalControllers) Save() in state/externalcontroller.go

Reply:
yup, changed